### ISSUE TITLE
`PATCH_COMMAND` via `FILE` manipulation doesn't seem to get cached

### ISSUE THREAD

**Issue by vittorioromeo** [cpm-cmake/CPM.cmake#586]
I'm trying to convert some `FetchContent` usages to CPM, but I am having trouble with existing `PATCH_COMMAND` arguments. Consider the following example:

```cmake
CPMAddPackage(
    NAME vorbis
    GITHUB_REPOSITORY xiph/vorbis
    GIT_TAG v1.3.7
    PATCH_COMMAND ${CMAKE_COMMAND} -DVORBIS_DIR=${FETCHCONTENT_BASE_DIR}/vorbis-src -P ${CMAKE_CURRENT_LIST_DIR}/../../../tools/vorbis/PatchVorbis.cmake)
```

`PatchVorbis.cmake` is as follows:

```cmake
file(READ "${VORBIS_DIR}/CMakeLists.txt" VORBIS_CMAKELISTS_CONTENTS)
string(REPLACE "VERSION 2.8.12" "VERSION 3.24" VORBIS_CMAKELISTS_CONTENTS "${VORBIS_CMAKELISTS_CONTENTS}")
# ...more replacements...
file(WRITE "${VORBIS_DIR}/CMakeLists.txt" "${VORBIS_CMAKELISTS_CONTENTS}")
```

I can see that the patch is applied to `./<build-folder>/_deps/vorbis-src/CMakeLists.txt`, but it is not applied to the `CMakeLists.txt` in the `$CPM_SOURCE_CACHE` location.

What is the proper way of solving this problem?

**Comment by ScottBailey** [comment]
First, `PATCH_COMMAND` is a pass through to CMake FetchContent. Not that this impacts your question. 

I think you wan to avoid referencing `FETCHCONTENT_BASE_DIR` and focus on where CPM puts your source. Try looking at `vorbis_SOURCE_DIR` instead. 

i.e. use this: `-DVORBIS_DIR=${vorbis_SOURCE_DIR}`.

Also, it might be more maintainable _in the long run_ to use `diff -ruN` and the `PATCHES` keyword instead of using a custom script.  If you go this route, you'll probably want to make sure you have CPM_SOURCE_CACHE set. Eventually CMake will have high quality patching, until then PATCHES is an okay work around.

**Comment by vittorioromeo** [comment]
> i.e. use this: `-DVORBIS_DIR=${vorbis_SOURCE_DIR}`.

That doesn't seem to work -- that variable seems to be empty during the invocation of `CPMAddPackage` and the `PATCH_COMMAND`, but it does get filled up right afterwards. I guess `PATCH_COMMAND` gets expanded before `vorbis_SOURCE_DIR` is set...?

**Comment by ScottBailey** [comment]
It kind of looks like you'll need to manage this by hand or generate a static patch.

**Comment by vittorioromeo** [comment]
> It kind of looks like you'll need to manage this by hand or generate a static patch.

Isn't this use case something that CPM is willing to support?

The reason I've chosen to use CPM is that it's basically a drop-in replacement for FetchContent that supports caching.

There are some instances where a static patch doesn't cut it, and `PATCH_COMMAND` is the way to go for traditional FetchContent.

Wouldn't it be reasonable for CPM to either support `PATCH_COMMAND` directly or to provide a way of injecting the cache directory after the download occurs but before the downloaded project is included?

**Comment by ScottBailey** [comment]
Let me preface the rest of my reply with this: I am _not_ a CPM maintainer, I'm just a dumb engineer that thinks CPM a cool project that I want to use everywhere so I'm going to provide help where I can. I'm guessing your use case is more of a corner case and isn't likely to merit the kind of attention that some of the other existing issues have.

Still, it's a valid problem for  you and feels like it could be fixed (or at least better understood) without too much pain. Let's you and me fix it. I started looking at the CPM code last night, but nothing jumper out at me so I took a look again this AM. But I only have a little time.


MAYBE we are overthinking this. Wouldn't you imagine that this should work if a basic simple patch command works? I would. Yeah, I bet we are.


Try this: get rid of the directory command in your patching call. i.e.:
```
    PATCH_COMMAND ${CMAKE_COMMAND} -P ${CMAKE_CURRENT_LIST_DIR}/../../../tools/vorbis/PatchVorbis.cmake)
```

I bet it just works.

**Comment by vittorioromeo** [comment]
@ScottBailey: that doesn't seem to work either, because `PatchVorbis.cmake` is executed in a completely separate `cmake` process, and has no access to the variables/scope of the main `cmake` process. I managed to find a workaround, but it's really horrible:

```cmake
CPMAddPackage(
    NAME vorbis
    GITHUB_REPOSITORY xiph/vorbis
    GIT_TAG v1.3.7
    DOWNLOAD_ONLY YES # <==
    PATCH_COMMAND ""  # <== needed to suppress "dirty cache CPM warning"
)

execute_process(COMMAND ${CMAKE_COMMAND} -DVORBIS_DIR=${vorbis_SOURCE_DIR} -P ${CMAKE_CURRENT_LIST_DIR}/../../../tools/vorbis/PatchVorbis.cmake)
FetchContent_MakeAvailable(vorbis)
add_subdirectory(${vorbis_SOURCE_DIR} ${vorbis_BINARY_DIR})
```

I'd really like to see support for this in CPM itself.

**Comment by ScottBailey** [comment]
I still suspect you are overthinking this as well as misunderstanding the intrinsics of CMake itself. Shoot, it's taken me some time to figure out what's wrong with your code and I'm gonna guess I've got a lot more expertise with both CMake and CPM than most people do. But I hope I can explain it well enough to become obvious to you. 

--------------------------

First, `PATCH_COMMAND` is a pass through to `FetcContent_Declare()` (as seen on line 1016 of CPM.cmake). The command is actually executed in what would be the source download directory.

Second, in the following code:
```
CPMAddPackage(
    NAME vorbis
    GITHUB_REPOSITORY xiph/vorbis
    GIT_TAG v1.3.7
    PATCH_COMMAND  ${CMAKE_COMMAND} -DVAR1=${ANYTHING} -P PatchVorbis.cmake
  )
```
`${ANYTHING}` is evaluated BEFORE  the function `CPMAddPackage` is called.

Changing this to do what you want it to is not possible without delaying the evaluation of ANYTHING. That would need to be done with CMake. Furthermore, it's actually an unexpected behavior.

------------------

BUT that doesn't mean we can't make this work!  First, I am not able to repeat the defect you are reporting. I got the following code to work without any workarounds.

```cmake
# CMakeLists.txt
cmake_minimum_required(VERSION 3.30)
project(test_stuff)

# Download CPM.cmake
file(
  DOWNLOAD      https://github.com/cpm-cmake/CPM.cmake/releases/download/v0.40.0/CPM.cmake ${CMAKE_CURRENT_BINARY_DIR}/cmake/CPM.cmake
  EXPECTED_HASH SHA256=7b354f3a5976c4626c876850c93944e52c83ec59a159ae5de5be7983f0e17a2a
)
include(${CMAKE_CURRENT_BINARY_DIR}/cmake/CPM.cmake)


CPMAddPackage(
    NAME vorbis
    GITHUB_REPOSITORY xiph/vorbis
    GIT_TAG v1.3.7
    PATCH_COMMAND ${CMAKE_COMMAND} -P ${CMAKE_CURRENT_LIST_DIR}/PatchVorbis.cmake
  )

add_executable(f main.cpp)
target_link_libraries(f vorbis)

```

```cmake
#PatchVorbis.cmake
file(READ "CMakeLists.txt" VORBIS_CMAKELISTS_CONTENTS)
string(REPLACE "VERSION 2.8.12" "VERSION 3.24" VORBIS_CMAKELISTS_CONTENTS "${VORBIS_CMAKELISTS_CONTENTS}")
# ...more replacements...
file(WRITE "CMakeLists.txt" "${VORBIS_CMAKELISTS_CONTENTS}")
```

I confirmed that CMakeLists.txt was patched in the file `~/.cache/CPM/vorbis/23ccbbe5c2ed702027c4a2ccf6f8a8cb123d29f8/CMakeLists.txt`.

--------------------

Frankly, I don't know why you would, but if you really need to set VORBIS_DIR you can set it in your PatchVorbis file : `set(VORBIS_DIR "${CMAKE_CURRENT_BINARY_DIR}")`

At the point PatchVorbis.cmake is called, `CMAKE_CURRENT_LIST_DIR` is as you'd expect (the path containing PatchVorbis.cmake) and `CMAKE_CURRENT_BINARY_DIR` is the path containing the downloaded source (for me this was `/home/bailey/.cache/CPM/vorbis/23ccbbe5c2ed702027c4a2ccf6f8a8cb123d29f8`)

-------------------------
TLDR the change request as stated is impossible from within CPM, it requires changes to CMake proper. I don't think those changes would be approved because of the resulting inconsistencies in behavior.

**Comment by ScottBailey** [comment]
@vittorioromeo Does this answer the issue for you? If so, please close. Thanks!

**Comment by vittorioromeo** [comment]
> @vittorioromeo Does this answer the issue for you? If so, please close. Thanks!

Sorry for the late reply, your solution did ended up working for me. I'm closing this issue right noow, but it would be nice to see support for `PATCH_COMMAND` in CPM.
